### PR TITLE
borrowck: consolidate `mut` suggestions

### LINKED ISSUE #35937: "cannot assign to immutable field" could be confusing
``` rust
struct S { x: i32 }
let s = S { x: 42 };
s.x += 1; //~ERROR cannot assign to immutable field
```

It occurred to me that this might be confusing to newcomers because fields don't have mutability in Rust, instead we have inherited mutability. So maybe it should say "cannot assign to field of immutable struct" or "cannot modify field due to immutable binding" or something.

cf #18150

### LINKED ISSUE #40823: 1.17 nightly Compiler suggests wrong syntax `mut mut var`
When trying to compile

```rust
fn main() {
    let mut buf = &[1, 2, 3, 4];
    let mut dst = &[0, 0, 0, 0];
    for (x, y) in buf.iter_mut().zip(dst) {
        *y = *x
    }
}
```

the compiler rustc 1.17.0-nightly (3da40237e 2017-03-24) throws an error together with a helpful suggestion

```
error: cannot borrow immutable borrowed content `*buf` as mutable
 --> <anon>:4:19
  |
2 |     let mut buf = &[1, 2, 3, 4];
  |         ------- consider changing this to `mut mut buf`
3 |     let mut dst = &[0, 0, 0, 0];
4 |     for (x, y) in buf.iter_mut().zip(dst) {
  |                   ^^^ cannot borrow as mutable
```

Here, the `mut mut buf` is obviously very very wrong. The stable and beta compilers don't suffer from this issue (they just provide no „consider …“ message at all).

### LINKED ISSUE #40859: Unparseable type suggested in error message
```
fn m(var: &[&mut u8]) {
    for i in var.iter_mut() {
        **i = 4;
    }
}

fn main() {
    let mut a = 1;
    let mut b = 2;
    let mut c = 3;
    let mut v = vec![&mut a, &mut b, &mut c];
    
    m(&mut v);
    println!("{:?}", v);
}
```

Gives error:

```
rustc 1.17.0-nightly (7846dbe0c 2017-03-26)
error: cannot borrow immutable borrowed content `*var` as mutable
 --> <anon>:2:14
  |
1 | fn m(var: &[&mut u8]) {
  |           ---------- use `&mut [&mut mut u8]` here to make mutable
2 |     for i in var.iter_mut() {
  |              ^^^ cannot borrow as mutable

error: aborting due to previous error
```

It suggests the type ```&mut [&mut mut u8]``` but that is not a valid Rust type. Happens in stable, beta, nightly.

### PR DESCRIPTION
**Pull request by arielb1**
This converts all of borrowck's `mut` suggestions to a new
`mc::ImmutabilityBlame` API instead of the current mix of various hacks.

Fixes #35937.
Fixes #40823.
Fixes #40859.

cc @estebank 
r? @pnkfelix

### CODE REVIEW

File: src/test/ui/did_you_mean/issue-39544.rs [hunk 1]
```diff
@@ -8,15 +8,20 @@
...
     Y
 }
 
-struct Z {
+pub struct Z {
     x: X
 }
 
-fn main() {
+pub fn main() {
```
Comment (estebank) [inline src/test/ui/did_you_mean/issue-39544.rs]: `pub fn main`?

File: src/librustc_borrowck/borrowck/mod.rs [hunk 2]
```diff
@@ -873,21 +837,93 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
...
+
+    fn local_ty(&self, node_id: ast::NodeId) -> Option<&hir::Ty> {
+        let parent = self.tcx.hir.get_parent_node(node_id);
         let parent_node = self.tcx.hir.get(parent);
 
         // The parent node is like a fn
         if let Some(fn_like) = FnLikeNode::from_node(parent_node) {
             // `nid`'s parent's `Body`
             let fn_body = self.tcx.hir.body(fn_like.body());
             // Get the position of `nid` in the arguments list
```
Comment (estebank) [inline src/librustc_borrowck/borrowck/mod.rs]: update comment to `node_id`

File: src/librustc_borrowck/borrowck/mod.rs [hunk 3]
```diff
@@ -873,21 +837,93 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
...
+        match blame {
+            None => {}
+            Some(ImmutabilityBlame::ClosureEnv(_)) => {}
+            Some(ImmutabilityBlame::ImmLocal(node_id)) => {
+                let let_span = self.tcx.hir.span(node_id);
+                if let hir::BindingMode::BindByValue(..) = self.local_binding_mode(node_id) {
+                    if let Ok(snippet) = self.tcx.sess.codemap().span_to_snippet(let_span) {
+                        if self.tcx.hir.name(node_id) == keywords::SelfValue.name() &&
+                            snippet != "self" {
+                            // avoid suggesting `mut &self`.
```
Comment (estebank) [inline src/librustc_borrowck/borrowck/mod.rs]: Shouldn't it suggest `&mut self` in that case?

File: src/librustc_borrowck/diagnostics.rs [hunk 4]
```diff
@@ -198,7 +198,7 @@ fn main() {
 ```
 "##,
 
-E0386: r##"
+/*E0386: r##"
```
Comment (estebank) [inline src/librustc_borrowck/diagnostics.rs]: What's the procedure to remove Errors that we no longer emit?

**Review by estebank — APPROVED** [review]
I like it (besides a couple of nitpicks).

**Comment by whataloadofwhat** [comment]
Is #35927 supposed to link somewhere else?

File: src/librustc_borrowck/diagnostics.rs (hunk 4 above)
Comment (arielb1) [inline src/librustc_borrowck/diagnostics.rs]: there's no such procedure.

File: src/librustc_borrowck/borrowck/mod.rs (hunk 3 above)
Comment (arielb1) [inline src/librustc_borrowck/borrowck/mod.rs]: Nope. Binding vs. reference - the code is intentionally careful to do the right thing.

**Comment by arielb1** [comment]
@whataloadofwhat 

Indeed. Updated & fixed.

BTW, @eddyb/@petrochenkov - is there a cleaner way of determining whether a function has an implicit self?

**Comment by eddyb** [comment]
@arielb1 Doubt it. It might be better idea to preserve the distinction in `hir::Ty` instead, not sure.

File: src/librustc_borrowck/borrowck/mod.rs (hunk 3 above)
Comment (estebank) [inline src/librustc_borrowck/borrowck/mod.rs]: I see, you're right, of course.

File: src/librustc_borrowck/diagnostics.rs (hunk 4 above)
Comment (nagisa) [inline src/librustc_borrowck/diagnostics.rs]: We put these into `register_diagnostics! {}`, commented out, although a better way to annotate codes that may not be reused should be made.

Comment (pnkfelix) [inline src/librustc_borrowck/diagnostics.rs]: @nagisa are you saying that this modification (the commenting out here) is okay as is? Or are you saying it shouldn't stay here under `register_long_diagnostics!`, and should ... move to `register_diagnosics!`...?

I think you must mean the former, that this is okay as is... Let us know if I misunderstand.

Comment (pnkfelix) [inline src/librustc_borrowck/diagnostics.rs]: (well I guess deleting all this long commented-out diagnostic text and replacing it with one line in `register_diagnostics!` could be an attractive option...)

**Pull request merged**